[snap][testing] Fix bionic snap runtime

arif-ali opened the pull request:
When the snap is being packaged via core22, it used python3.10
but bionic used python3.6 and seems have runtime issues there.

Added the sos alias for testing, to ensure that when the stage
one testing is done, that sos is correctly tested within the
snap

This alternative ensures that the packaging module is packaged
in the snap, and can use this if the pkg_resources version does
not work

Add a wrapper function in utlities, so that the parse_version
can be used in a consistent way across the sos project

Closes: #3126

Signed-off-by: Arif Ali <arif.ali@canonical.com>

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [x] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?

arif-ali commented:
merging #3128 into this as it's just one-liner and 2 combined should result in the key things being resolved here

arif-ali commented:
Okie, more 18.04 issues, needs further debugging/analysis

TurboTurtle commented:
The error message seems the same, perhaps the catch block is too specific and we should just have a generic `except Exception` on any failure to import from `pkg_resources`

arif-ali commented:
It's not that trivial it seems, there's no stack trace

```
root@bionic02:~# export SNAP=/snap/sosreport/current
root@bionic02:~# export PYTHONPATH=${SNAP}/usr/lib/python3.10:${SNAP}/lib/python3.10/site-packages:${SNAP}/usr/lib/python3/dist-packages
root@bionic02:~# python3 -m pdb  /snap/sosreport/current/bin/sos collect --list-options
> /snap/sosreport/current/bin/sos(10)<module>()
-> import sys
(Pdb) c
Could not initialize 'collect': future feature annotations is not defined (__init__.py, line 5)
The program exited via sys.exit(). Exit status: 1
> /snap/sosreport/current/bin/sos(10)<module>()
-> import sys
(Pdb) 
root@bionic02:~# sos collect
Could not initialize 'collect': future feature annotations is not defined (__init__.py, line 5)
```

I don't think it's a pkg_resource issue now, as `sos report` works, it's an issue with `sos collect`

TurboTurtle commented:
I don't follow how it's an issue in `sos collect`. The full error in #3126 is clear:

```
  File "/snap/sosreport/70/lib/python3.10/site-packages/sos/collector/sosnode.py", line 17, in <module>
    from pkg_resources import parse_version
  File "/snap/sosreport/70/lib/python3.10/site-packages/pkg_resources/__init__.py", line 82, in <module>
    from pkg_resources.extern import platformdirs
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 658, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 571, in module_from_spec
  File "/snap/sosreport/70/lib/python3.10/site-packages/pkg_resources/extern/__init__.py", line 52, in create_module
    return self.load_module(spec.name)
  File "/snap/sosreport/70/lib/python3.10/site-packages/pkg_resources/extern/__init__.py", line 37, in load_module
    __import__(extant)
  File "/snap/sosreport/70/lib/python3.10/site-packages/pkg_resources/_vendor/platformdirs/__init__.py", line 5
    from __future__ import annotations
    ^
SyntaxError: future feature annotations is not defined
```

`annotations` is available from python-3.7+, this error implies that the python-3.6 interpreter from the bionic host is trying to import a python-3.10 `pkg_resources` module from the snap installation.

You're seeing this on any invocation of `collect` because `collect` imports `SoSNode` which in turn imports from `pkg_resources` which calls this future import.

arif-ali commented:
* I've added `except Exception` to test locally, and we have the same issue
* I've removed the whole line of of pkg_resources, and just use the packaging module, and still have the same issue

I'll keep debugging :)

arif-ali commented:
In the end it was another python files, and found the issue by running `sos collect --verbose --debug`, and it was the pacemaker.py in the collector (below the Traceback for reference)

```
$ sos collect --debug --verbose
Could not initialize 'collect': future feature annotations is not defined (__init__.py, line 5)
Traceback (most recent call last):
  File "./sos/bin/sos", line 21, in <module>
    sos = SoS(sys.argv[1:])
  File "/snap/sosreport/current/lib/python3.10/site-packages/sos/__init__.py", line 124, in __init__
    self._init_component()
  File "/snap/sosreport/current/lib/python3.10/site-packages/sos/__init__.py", line 189, in _init_component
    raise err
  File "/snap/sosreport/current/lib/python3.10/site-packages/sos/__init__.py", line 184, in _init_component
    self._component = _to_load(self.parser, self.args, self.cmdline)
  File "/snap/sosreport/current/lib/python3.10/site-packages/sos/collector/__init__.py", line 174, in __init__
    self.clusters = self.load_clusters()
  File "/snap/sosreport/current/lib/python3.10/site-packages/sos/collector/__init__.py", line 197, in load_clusters
    clusters = self._load_modules(package, 'clusters')
  File "/snap/sosreport/current/lib/python3.10/site-packages/sos/collector/__init__.py", line 208, in _load_modules
    modules.extend(cls._find_modules_in_path(path, submod))
  File "/snap/sosreport/current/lib/python3.10/site-packages/sos/collector/__init__.py", line 231, in _find_modules_in_path
    modules.extend(cls._import_modules(modname))
  File "/snap/sosreport/current/lib/python3.10/site-packages/sos/collector/__init__.py", line 238, in _import_modules
    module = __import__(modname, globals(), locals(), [mod_short_name])
  File "/snap/sosreport/current/lib/python3.10/site-packages/sos/collector/clusters/pacemaker.py", line 14, in <module>
    from setuptools._vendor.packaging import version
  File "/snap/sosreport/current/lib/python3.10/site-packages/setuptools/__init__.py", line 16, in <module>
    import setuptools.version
  File "/snap/sosreport/current/lib/python3.10/site-packages/setuptools/version.py", line 1, in <module>
    import pkg_resources
  File "/snap/sosreport/current/lib/python3.10/site-packages/pkg_resources/__init__.py", line 82, in <module>
    from pkg_resources.extern import platformdirs
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 658, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 571, in module_from_spec
  File "/snap/sosreport/current/lib/python3.10/site-packages/pkg_resources/extern/__init__.py", line 52, in create_module
    return self.load_module(spec.name)
  File "/snap/sosreport/current/lib/python3.10/site-packages/pkg_resources/extern/__init__.py", line 37, in load_module
    __import__(extant)
  File "/snap/sosreport/current/lib/python3.10/site-packages/pkg_resources/_vendor/platformdirs/__init__.py", line 5
    from __future__ import annotations
    ^
SyntaxError: future feature annotations is not defined
```

TurboTurtle commented on sos/collector/clusters/pacemaker.py:
```diff
@@ -11,7 +11,10 @@
 
 from sos.collector.clusters import Cluster
-from setuptools._vendor.packaging import version
+try:
+    from setuptools._vendor.packaging import version
+except SyntaxError:
+    from packaging import version
```
Nice find.

Looking at this, we may want to drop the `from setuptools._vendor` import, and instead mirror the use of `parse_version` like we're (now) doing in `sosnode`. This would give us some consistency at least.

The `pacemaker` check here is a simple one-off check to handle how an option is formatted. So changing to something like:

```diff
@@ -11,7 +11,7 @@
 import re
 
 from sos.collector.clusters import Cluster
-from setuptools._vendor.packaging import version
+from pkg_resources import parse_version
 from xml.etree import ElementTree
 
 
@@ -63,7 +63,7 @@ class pacemaker(Cluster):
         _ver = self.exec_primary_cmd('crm_mon --version')
         if _ver['status'] == 0:
             cver = _ver['output'].split()[1].split('-')[0]
-            if not version.parse(cver) > version.parse('2.0.3'):
+            if not parse_version(cver) > parse_version('2.0.3'):
                 xmlopt = '--as-xml'
         else:
             return
```

Would be an easy replacement.

Alternatively, if we're going to be doing this in multiple spots, we could simply add a wrapper in `utilities.py`.

arif-ali replied:
ack, will make it consistent, just waiting for all the tests to pass, and will push an update

arif-ali commented:
decided to the wrapper way, looks nicer, and can be re-used
